### PR TITLE
Noop on build

### PR DESCRIPTION
**Pull request by emgarten**
This change validates the lock files upfront and verifies that all packages exist to avoid showing the progress dialog and performing full restores which are unneeded for scenarios such as building twice in a row with no changes.

https://github.com/NuGet/Home/issues/845
https://github.com/NuGet/Home/issues/844

### CODE REVIEW

**Comment by emgarten** [comment]
//cc @yishaigalatzer @deepakaravindr

File: src/VsExtension/VsEvents/OnBuildPackageRestorer.cs [hunk 1]
```diff
@@ -33,6 +34,9 @@ internal sealed class OnBuildPackageRestorer
         private readonly DTE _dte;
         private bool _outputOptOutMessage;
 
+        private IReadOnlyDictionary<string, BuildIntegratedProjectCacheEntry> _buildIntegratedCache
```
Comment (yishaigalatzer) [inline src/VsExtension/VsEvents/OnBuildPackageRestorer.cs:37]: readonly

File: src/VsExtension/VsEvents/OnBuildPackageRestorer.cs [hunk 2]
```diff
@@ -162,7 +166,10 @@ private void BuildEvents_OnBuildBegin(vsBuildScope scope, vsBuildAction Action)
                         // Call DNU to restore for BuildIntegratedProjectSystem projects
                         var buildEnabledProjects = projects.OfType<BuildIntegratedProjectSystem>();
 
-                        await RestoreBuildIntegratedProjectsAsync(buildEnabledProjects);
+                        var forceRestore = Action == vsBuildAction.vsBuildActionRebuildAll
```
Comment (yishaigalatzer) [inline src/VsExtension/VsEvents/OnBuildPackageRestorer.cs:169]: is vsbuildaction guaranteed not to be null?

File: src/VsExtension/VsEvents/OnBuildPackageRestorer.cs (hunk 1 above)
Comment (emgarten) [inline src/VsExtension/VsEvents/OnBuildPackageRestorer.cs:37]: the cache is swapped out each time, it gets the current closure, compares, and replaces it for the next time

File: src/VsExtension/VsEvents/OnBuildPackageRestorer.cs [hunk 3]
```diff
@@ -187,7 +194,7 @@ private void BuildEvents_OnBuildBegin(vsBuildScope scope, vsBuildAction Action)
             }
         }
 
-        private async Task RestoreBuildIntegratedProjectsAsync(IEnumerable<BuildIntegratedProjectSystem> buildEnabledProjects)
+        private async Task RestoreBuildIntegratedProjectsAsync(List<BuildIntegratedProjectSystem> buildEnabledProjects, bool forceRestore)
```
Comment (yishaigalatzer) [inline src/VsExtension/VsEvents/OnBuildPackageRestorer.cs:197]: break the long line, and add a comment that force restore is used when a rebuild is called

File: src/VsExtension/VsEvents/OnBuildPackageRestorer.cs [hunk 4]
```diff
@@ -201,42 +208,87 @@ private void BuildEvents_OnBuildBegin(vsBuildScope scope, vsBuildAction Action)
...
+                    if (forceRestore || await IsRestoreRequired(buildEnabledProjects))
                     {
+                        var waitDialogFactory = ServiceLocator.GetGlobalService<SVsThreadedWaitDialogFactory, IVsThreadedWaitDialogFactory>();
+
                         // NOTE: During restore for build integrated projects,
-                        //       We might show PackageRestoreOptOutMessage even if there are no packages to restore
-                        WriteLine(VerbosityLevel.Quiet, Resources.PackageRestoreOptOutMessage);
-                        _outputOptOutMessage = false;
+                        //       We might show the dialog even if there are no packages to restore
+                        // When both currentStep and totalSteps are 0, we get a marquee on the dialog
```
Comment (yishaigalatzer) [inline src/VsExtension/VsEvents/OnBuildPackageRestorer.cs:218]: fix the spacing

File: src/VsExtension/VsEvents/OnBuildPackageRestorer.cs [hunk 5]
```diff
@@ -201,42 +208,87 @@ private void BuildEvents_OnBuildBegin(vsBuildScope scope, vsBuildAction Action)
...
-                        {
-                            var projectName = NuGetProject.GetUniqueNameOrName(project);
-                            await BuildIntegratedProjectRestoreAsync(project, loggingProjectContext, enabledSources, Token);
-                            WriteLine(VerbosityLevel.Normal, Resources.PackageRestoreFinishedForProject, projectName);
-                        }
+                            // Restore packages and create the lock file for each project
+                            foreach (var project in buildEnabledProjects)
+                            {
+                                var projectName = NuGetProject.GetUniqueNameOrName(project);
+                                await BuildIntegratedProjectRestoreAsync(project, loggingProjectContext, enabledSources, Token);
```
Comment (yishaigalatzer) [inline src/VsExtension/VsEvents/OnBuildPackageRestorer.cs:240]: break the long lines

File: src/VsExtension/VsEvents/OnBuildPackageRestorer.cs [hunk 6]
```diff
@@ -201,42 +208,87 @@ private void BuildEvents_OnBuildBegin(vsBuildScope scope, vsBuildAction Action)
...
+
+                if (BuildIntegratedRestoreUtility.IsRestoreRequired(projects, pathResolver))
+                {
+                    // The project.json file does not match the lock file
+                    return true;
+                }
+            }
+            catch
+            {
+                Debug.Fail("Unable to validate lock files.");
```
Comment (yishaigalatzer) [inline src/VsExtension/VsEvents/OnBuildPackageRestorer.cs:287]: add the exception details to the failure, and consider logging or tracing something in this case so if it happens repeatadly in production we can ask for data from the customer

File: src/VsExtension/VsEvents/OnBuildPackageRestorer.cs [hunk 7]
```diff
@@ -201,42 +208,93 @@ private void BuildEvents_OnBuildBegin(vsBuildScope scope, vsBuildAction Action)
...
+        /// where all packages are already on disk and no server requests are needed
+        /// we can skip the restore after some validation checks.
+        /// </summary>
+        private async Task<bool> IsRestoreRequired(List<BuildIntegratedProjectSystem> projects, bool forceRestore)
+        {
+            try
+            {
+                // Swap caches 
+                var oldCache = _buildIntegratedCache;
+                _buildIntegratedCache = await BuildIntegratedRestoreUtility.CreateBuildIntegratedProjectStateCache(projects);
```
Comment (deepakaravindr) [inline src/VsExtension/VsEvents/OnBuildPackageRestorer.cs:262]: Break the long line

**Comment by deepakaravindr** [comment]
:shipit:

File: src/VsExtension/VsEvents/OnBuildPackageRestorer.cs (hunk 6 above)
Comment (emgarten) [inline src/VsExtension/VsEvents/OnBuildPackageRestorer.cs:287]: I've added writing to the output under diagnostic mode, and logging to the activity log for this.

File: src/VsExtension/VsEvents/OnBuildPackageRestorer.cs (hunk 2 above)
Comment (emgarten) [inline src/VsExtension/VsEvents/OnBuildPackageRestorer.cs:169]: it's an enum and cannot be null

**Pull request closed without merging**